src/xenia/cpu/processor.cc

```diff
@@ -383,5 +405,74 @@ void Processor::LowerIrql(Irql old_value) {
                       reinterpret_cast<volatile uint32_t*>(&irql_));
 }
 
+void Processor::BreakpointFunctionDefined(Function* function) {
+  std::lock_guard<std::recursive_mutex> lock(breakpoint_lock_);
+
+  for (auto it = breakpoints_.begin(); it != breakpoints_.end(); it++) {
+    if (function->ContainsAddress((*it)->address())) {
+      backend_->InstallBreakpoint(*it, function);
+    }
+  }
+}
+
+bool Processor::InstallBreakpoint(Breakpoint* bp) {
+  std::lock_guard<std::recursive_mutex> lock(breakpoint_lock_);
+
+  if (FindBreakpoint(bp->address())) {
+    return false;
+  }
+
+  // We need to register the breakpoint before installing it with the backend
+  // in-case a thread hits it while we're here.
+  breakpoints_.push_back(bp);
+  if (!backend_->InstallBreakpoint(bp)) {
+    breakpoints_.pop_back();
+    return false;
+  }
+
+  return true;
+}
+
+bool Processor::UninstallBreakpoint(Breakpoint* bp) {
+  std::lock_guard<std::recursive_mutex> lock(breakpoint_lock_);
+
+  if (!backend_->UninstallBreakpoint(bp)) {
+    return false;
+  }
+
+  for (auto it = breakpoints_.begin(); it != breakpoints_.end(); it++) {
```

Comment:
would it not matter if bp was a copy or reference if the value is just a pointer to a breakpoint? I might be mistaken though (probably am)